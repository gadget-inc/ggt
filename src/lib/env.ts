import type { Paths } from "env-paths";
import envPaths from "env-paths";

/**
 * Captures the name and nature of the environment
 */
export class Env {
  static get value(): string {
    return process.env["GGT_ENV"] || "production";
  }

  static get paths(): Paths {
    return envPaths("gadget", { suffix: this.productionLike ? "" : this.value });
  }

  static get developmentLike(): boolean {
    return this.value.startsWith("development");
  }

  static get testLike(): boolean {
    return this.value.startsWith("test");
  }

  static get developmentOrTestLike(): boolean {
    return this.developmentLike || this.testLike;
  }

  static get productionLike(): boolean {
    return this.value.startsWith("production");
  }
}
