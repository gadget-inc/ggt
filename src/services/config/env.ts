import process from "node:process";

/**
 * Captures the name and nature of the environment
 */
export const env = {
  get value(): string {
    return process.env["GGT_ENV"] || "production";
  },

  get productionLike(): boolean {
    return !this.developmentOrTestLike;
  },

  get developmentLike(): boolean {
    return this.value.startsWith("development");
  },

  get testLike(): boolean {
    return this.value.startsWith("test");
  },

  get developmentOrTestLike(): boolean {
    return this.developmentLike || this.testLike;
  },
};
