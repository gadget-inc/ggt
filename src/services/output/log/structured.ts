import { addBreadcrumb as addSentryBreadcrumb } from "@sentry/node";
import { config } from "../../config/config.js";
import { env } from "../../config/env.js";
import { unthunk, type Thunk } from "../../util/function.js";
import { serializeError } from "../../util/object.js";
import { stderr } from "../stream.js";
import type { Fields } from "./field.js";
import { formatters } from "./format/format.js";
import { Level } from "./level.js";

type StructuredLog = (msg: Lowercase<string>, fields?: Fields, devFields?: Fields) => void;

export type StructuredLogger = {
  trace: StructuredLog;
  debug: StructuredLog;
  info: StructuredLog;
  warn: StructuredLog;
  error: StructuredLog;
};

export type StructuredLoggerOptions = {
  /**
   * The name of logger.
   */
  name: string;

  /**
   * Fields to add to every message logged by the logger.
   */
  fields?: Thunk<Fields>;

  /**
   * Fields to add to every message logged by the logger only in
   * development or test environments.
   */
  devFields?: Thunk<Fields>;
};

export const createStructuredLogger = ({
  name,
  fields: loggerFields = {},
  devFields: devLoggerFields = {},
}: StructuredLoggerOptions): StructuredLogger => {
  const createStructuredLog = (level: Level): StructuredLog => {
    return (msg, messageFields, devMessageFields) => {
      const shouldLog = level >= config.logLevel;
      const shouldSendToSentry = level >= Level.INFO;
      if (!shouldLog && !shouldSendToSentry) {
        return;
      }

      const fields = { ...unthunk(loggerFields), ...messageFields };
      if (env.developmentOrTestLike) {
        Object.assign(fields, unthunk(devLoggerFields), devMessageFields);
      }

      if ("error" in fields) {
        fields.error = serializeError(fields.error);
      }

      if ("reason" in fields) {
        fields.reason = serializeError(fields.reason);
      }

      if (shouldLog) {
        const format = formatters[config.logFormat];
        stderr.write(format(level, name, msg, fields));
      }

      if (shouldSendToSentry) {
        addSentryBreadcrumb({ level: "log", message: msg, data: fields });
      }
    };
  };

  return {
    trace: createStructuredLog(Level.TRACE),
    debug: createStructuredLog(Level.DEBUG),
    info: createStructuredLog(Level.INFO),
    warn: createStructuredLog(Level.WARN),
    error: createStructuredLog(Level.ERROR),
  };
};
