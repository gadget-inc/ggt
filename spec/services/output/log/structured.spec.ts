import * as Sentry from "@sentry/node";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { createStructuredLogger, setGlobalFields } from "../../../../src/services/output/log/structured.js";
import { withEnv } from "../../../__support__/env.js";
import { expectStderr, mockStderr } from "../../../__support__/stdout.js";

describe("structured", () => {
  mockStderr();

  beforeEach(() => {
    setGlobalFields({});
    vi.useFakeTimers();
    vi.setSystemTime(0);
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it.each(["trace", "debug", "info", "warn", "error"])("logs the expected output when GGT_LOG_LEVEL=%s", (level) => {
    const structuredLogger = createStructuredLogger({ name: "structured" });

    withEnv({ GGT_LOG_LEVEL: level }, () => {
      structuredLogger.trace("trace");
      structuredLogger.debug("debug");
      structuredLogger.info("info");
      structuredLogger.warn("warn");
      structuredLogger.error("error");
    });

    expectStderr().toMatchSnapshot();
  });

  it("adds the expected logs to sentry breadcrumbs", () => {
    vi.spyOn(Sentry, "addBreadcrumb");

    const structuredLogger = createStructuredLogger({ name: "structured" });

    withEnv({ GGT_LOG_LEVEL: "trace" }, () => {
      structuredLogger.trace("trace");
      structuredLogger.debug("debug");
      structuredLogger.info("info");
      structuredLogger.warn("warn");
      structuredLogger.error("error");
    });

    expect(Sentry.addBreadcrumb).toHaveBeenCalledTimes(3);
    expect(Sentry.addBreadcrumb).toHaveBeenCalledWith({ level: "log", message: "info", data: {} });
    expect(Sentry.addBreadcrumb).toHaveBeenCalledWith({ level: "log", message: "warn", data: {} });
    expect(Sentry.addBreadcrumb).toHaveBeenCalledWith({ level: "log", message: "error", data: {} });
  });

  it("merges global, logger, and message fields", () => {
    withEnv({ GGT_LOG_LEVEL: "trace" }, () => {
      let structuredLogger = createStructuredLogger({ name: "structured" });

      setGlobalFields({ global: true, field: "global" });
      structuredLogger.trace("first message");
      expectStderr().toMatchInlineSnapshot(`
        "12:00:00 TRACE structured: first message
          global: true
          field: 'global'
        "
      `);

      structuredLogger = createStructuredLogger({ name: "structured", fields: { logger: true, field: "logger" } });
      structuredLogger.trace("second message");
      expectStderr().toMatchInlineSnapshot(`
        "12:00:00 TRACE structured: first message
          global: true
          field: 'global'
        12:00:00 TRACE structured: second message
          global: true
          field: 'logger'
          logger: true
        "
      `);

      structuredLogger.trace("third message", { message: true, field: "message" });
      expectStderr().toMatchInlineSnapshot(`
        "12:00:00 TRACE structured: first message
          global: true
          field: 'global'
        12:00:00 TRACE structured: second message
          global: true
          field: 'logger'
          logger: true
        12:00:00 TRACE structured: third message
          global: true
          field: 'message'
          logger: true
          message: true
        "
      `);
    });
  });

  it.each(["development", "test"])("prints dev fields in %s", (name) => {
    const structuredLogger = createStructuredLogger({ name: "structured" });

    withEnv({ GGT_LOG_LEVEL: "trace", GGT_ENV: name }, () => {
      structuredLogger.trace("trace", {}, { dev: true });
      structuredLogger.debug("debug", {}, { dev: true });
      structuredLogger.info("info", {}, { dev: true });
      structuredLogger.warn("warn", {}, { dev: true });
      structuredLogger.error("error", {}, { dev: true });
    });

    expectStderr().toMatchSnapshot();
  });

  it("does not print dev fields in production", () => {
    const structuredLogger = createStructuredLogger({ name: "structured" });

    withEnv({ GGT_LOG_LEVEL: "trace", GGT_ENV: "production" }, () => {
      structuredLogger.trace("trace", {}, { dev: true });
      structuredLogger.debug("debug", {}, { dev: true });
      structuredLogger.info("info", {}, { dev: true });
      structuredLogger.warn("warn", {}, { dev: true });
      structuredLogger.error("error", {}, { dev: true });
    });

    expectStderr().toMatchSnapshot();
  });

  it("prints json when GGT_LOG_FORMAT=json", () => {
    const structuredLogger = createStructuredLogger({ name: "structured" });

    withEnv({ GGT_LOG_LEVEL: "trace", GGT_LOG_FORMAT: "json" }, () => {
      structuredLogger.trace("trace");
      structuredLogger.debug("debug");
      structuredLogger.info("info");
      structuredLogger.warn("warn");
      structuredLogger.error("error");
    });

    expectStderr().toMatchSnapshot();
  });
});
