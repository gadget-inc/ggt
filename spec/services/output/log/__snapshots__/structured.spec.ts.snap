// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`structured > does not print dev fields in production 1`] = `
"12:00:00 TRACE structured: trace
12:00:00 DEBUG structured: debug
12:00:00 INFO structured: info
12:00:00 WARN structured: warn
12:00:00 ERROR structured: error
"
`;

exports[`structured > logs the expected output when GGT_LOG_LEVEL=debug 1`] = `
"12:00:00 DEBUG structured: debug
12:00:00 INFO structured: info
12:00:00 WARN structured: warn
12:00:00 ERROR structured: error
"
`;

exports[`structured > logs the expected output when GGT_LOG_LEVEL=error 1`] = `
"12:00:00 ERROR structured: error
"
`;

exports[`structured > logs the expected output when GGT_LOG_LEVEL=info 1`] = `
"12:00:00 INFO structured: info
12:00:00 WARN structured: warn
12:00:00 ERROR structured: error
"
`;

exports[`structured > logs the expected output when GGT_LOG_LEVEL=trace 1`] = `
"12:00:00 TRACE structured: trace
12:00:00 DEBUG structured: debug
12:00:00 INFO structured: info
12:00:00 WARN structured: warn
12:00:00 ERROR structured: error
"
`;

exports[`structured > logs the expected output when GGT_LOG_LEVEL=warn 1`] = `
"12:00:00 WARN structured: warn
12:00:00 ERROR structured: error
"
`;

exports[`structured > prints dev fields in development 1`] = `
"12:00:00 TRACE structured: trace
  dev: true
12:00:00 DEBUG structured: debug
  dev: true
12:00:00 INFO structured: info
  dev: true
12:00:00 WARN structured: warn
  dev: true
12:00:00 ERROR structured: error
  dev: true
"
`;

exports[`structured > prints dev fields in test 1`] = `
"12:00:00 TRACE structured: trace
  dev: true
12:00:00 DEBUG structured: debug
  dev: true
12:00:00 INFO structured: info
  dev: true
12:00:00 WARN structured: warn
  dev: true
12:00:00 ERROR structured: error
  dev: true
"
`;

exports[`structured > prints json when GGT_LOG_FORMAT=json 1`] = `
"{\\"level\\":1,\\"name\\":\\"structured\\",\\"msg\\":\\"trace\\",\\"fields\\":{}}
{\\"level\\":2,\\"name\\":\\"structured\\",\\"msg\\":\\"debug\\",\\"fields\\":{}}
{\\"level\\":3,\\"name\\":\\"structured\\",\\"msg\\":\\"info\\",\\"fields\\":{}}
{\\"level\\":4,\\"name\\":\\"structured\\",\\"msg\\":\\"warn\\",\\"fields\\":{}}
{\\"level\\":5,\\"name\\":\\"structured\\",\\"msg\\":\\"error\\",\\"fields\\":{}}
"
`;
